esp/services/esdl_svc_engine/esdl_binding.cpp

```diff
@@ -101,54 +101,151 @@ IPropertyTree * fetchESDLDefinitionFromDali(const char *esdlServiceName, unsigne
+        if (!conn)
+        {
+            DBGLOG("Unable to connect to ESDL Service binding information in dali %s", ESDL_BINDINGS_ROOT_PATH);
+            return NULL;
+        }
+
+        conn->close(false); //release lock right away
+
+        DBGLOG("ESDL Binding: Fetching ESDL Binding from Dali %s[@EspProcess='%s'][@EspBinding='%s'][1]", ESDL_BINDING_ENTRY, process, bindingName);
+
+        IPropertyTree *esdlBindings = conn->queryRoot();
+        if (!esdlBindings)
+        {
+            DBGLOG("Unable to open ESDL Service binding information in dali %s", ESDL_BINDINGS_ROOT_PATH);
+            return NULL;
+        }
+
+        //There shouldn't be multiple entries here, but if so, we'll use the first one
+        VStringBuffer xpath("%s[@id='%s.%s'][1]", ESDL_BINDING_ENTRY, process, bindingName);
+
+        return esdlBindings->getPropTree(xpath);
     }
+    catch(...)
+    {
+        DBGLOG("ESDL Binding: Unknown error encountered while fetching ESDL Binding from Dali %s[@EspProcess='%s'][@EspBinding='%s'][1]", ESDL_BINDING_ENTRY, process, bindingName);
+        if (conn)
+            conn->close(false); //release lock right away
 
-    conn->close(false); //release lock right away
+    }
+    return NULL;
+}
 
-    DBGLOG("ESDL Binding: Fetching ESDL Binding from Dali %s[@EspProcess='%s'][@EspBinding='%s'][1]", ESDL_BINDING_ENTRY, process, bindingName);
+IPropertyTree * fetchESDLBindingFromStateFile(const char *process, const char *bindingName, const char * stateFileName)
+{
+    Owned<IPTree> esdlState = createPTreeFromXMLFile(stateFileName);
+    if (!esdlState)
+        throw MakeStringException(-1, "Could not load DESDL binding state from local file: '%s' for ESP binding %s.%s", stateFileName, process, bindingName);
 
-    IPropertyTree *esdlBindings = conn->queryRoot();
-    if (!esdlBindings)
-       throw MakeStringException(-1, "Unable to open ESDL Service binding information in dali %s", ESDL_BINDINGS_ROOT_PATH);
+    const char * restoredBindingTS = esdlState->queryProp("@StateSaveTimems");
 
-    //There shouldn't be multiple entries here, but if so, we'll use the first one
-    VStringBuffer xpath("%s[@id='%s.%s'][1]", ESDL_BINDING_ENTRY, process, bindingName);
+    fprintf(stdout, "ESDL State restored from local state store: %s created epoch: %s", bindingName, restoredBindingTS);
 
-    return  LINK(esdlBindings->queryPropTree(xpath));
+    return esdlState->getPropTree("EsdlBinding/Binding");
+}
+
+IPropertyTree * fetchESDLBinding(const char *process, const char *bindingName, const char * stateFileName)
+{
+    Owned<IPropertyTree> esdlBinding = fetchESDLBindingFromDali(process, bindingName);
+    if (!esdlBinding)
+    {
+        return fetchESDLBindingFromStateFile(process, bindingName, stateFileName);
+    }
+
+    return esdlBinding.getLink();
+}
+
+void saveState(const char * esdlDefinition, IPropertyTree * esdlBinding, const char * stateFileFullPath)
```

Comment:
Static version of the saveState logic, callers are static. Added wrapper method to EsdlBinding class as well.